crates/ast/src/ast_node.rs

```diff
@@ -80,6 +80,14 @@ pub fn children_nodes<TNode: AstNode, TChildNode: AstNode>(
     node.syntax().children().filter_map(TChildNode::cast)
 }
 
+/// 指定したノードから指定した`kind`のトークンを探して最初の一つ目を返します。
+pub fn token(parent: &SyntaxNode, kind: SyntaxKind) -> Option<SyntaxToken> {
+    parent
+        .children_with_tokens()
+        .filter_map(|it| it.into_token())
+        .find(|it| it.kind() == kind)
+}
```

Comment:
**Optimization Issue**: The method 'token' retrieves the first token of a specified kind from a syntax node using a linear search, which can be inefficient for large syntax trees or frequent calls. <br> **Fix**: Implement caching for the results of token searches within syntax nodes to avoid repeated linear searches, or consider restructuring the data access patterns to reduce the need for such searches. <br> **Code Suggestion**: 
 ```
 +/// Cache for tokens to improve retrieval performance.
 +use std::collections::HashMap;
 +lazy_static! {
 +    static ref TOKEN_CACHE: Mutex<HashMap<SyntaxKind, Option<SyntaxToken>>> = Mutex::new(HashMap::new());
 +}
 +
 pub fn token(parent: &SyntaxNode, kind: SyntaxKind) -> Option<SyntaxToken> {
     let cache = TOKEN_CACHE.lock().unwrap();
 -    parent
 -        .children_with_tokens()
 -        .filter_map(|it| it.into_token())
 -        .find(|it| it.kind() == kind)
 +    if let Some(token) = cache.get(&kind) {
 +        token.clone()
 +    } else {
 +        let result = parent
 +            .children_with_tokens()
 +            .filter_map(|it| it.into_token())
 +            .find(|it| it.kind() == kind);
 +        cache.insert(kind, result.clone());
 +        result
 +    }
 +}
 ```